polar-core/src/parser.rs

```diff
@@ -31,75 +35,94 @@ pub enum Line {
     },
 }
 
-fn to_parse_error(e: ParseError<usize, lexer::Token, error::ParseError>) -> error::ParseError {
-    match e {
-        ParseError::InvalidToken { location: loc } => error::ParseError::InvalidToken { loc },
+fn lalrpop_error_to_polar_error(
+    e: ParseError<usize, lexer::Token, error::ParseErrorKind>,
+    source: &Arc<Source>,
+) -> error::ParseError {
+    let kind = match e {
+        ParseError::InvalidToken { location: loc } => error::ParseErrorKind::InvalidToken { loc },
         ParseError::UnrecognizedEOF { location: loc, .. } => {
-            error::ParseError::UnrecognizedEOF { loc }
+            error::ParseErrorKind::UnrecognizedEOF { loc }
         }
         ParseError::UnrecognizedToken {
             token: (loc, t, _), ..
         } => match t {
-            Token::Debug | Token::Cut | Token::In | Token::New => error::ParseError::ReservedWord {
-                token: t.to_string(),
-                loc,
-            },
-            _ => error::ParseError::UnrecognizedToken {
+            Token::Debug | Token::Cut | Token::In | Token::New => {
+                error::ParseErrorKind::ReservedWord {
+                    token: t.to_string(),
+                    loc,
+                }
+            }
+            _ => error::ParseErrorKind::UnrecognizedToken {
                 token: t.to_string(),
                 loc,
             },
         },
-        ParseError::ExtraToken { token: (loc, t, _) } => error::ParseError::ExtraToken {
+        ParseError::ExtraToken { token: (loc, t, _) } => error::ParseErrorKind::ExtraToken {
             token: t.to_string(),
             loc,
         },
         ParseError::User { error } => error,
+    };
+    error::ParseError {
+        source: source.clone(),
+        kind,
     }
 }
 
-pub fn parse_lines(src_id: u64, src: &str) -> Result<Vec<Line>, error::ParseError> {
+pub fn parse_lines(source: &Arc<Source>) -> Result<Vec<Line>, error::ParseError> {
```

Comment:
You might just be able to take `&Source` here. I think since arc impls `AsRef<Source>` (maybe getting this slightly wrong) you don't need the `Arc` part.

Reply:
Was taking `&Arc<Source>` here so that the same Arc would be cloned for [terms/rules generated by LALRPOP](https://github.com/osohq/oso/blob/3bd596971e476e9209652cece9df7aaff06481b2/polar-core/src/parser.rs#L75), [`ParseError`s](https://github.com/osohq/oso/blob/3bd596971e476e9209652cece9df7aaff06481b2/polar-core/src/parser.rs#L76), and [`FileLoading` errors](https://github.com/osohq/oso/blob/3bd596971e476e9209652cece9df7aaff06481b2/polar-core/src/polar.rs#L55). Was able to reduce the # of places we need to care about Arcs by [having `FileLoading` errors assemble their own `Source`](https://github.com/osohq/oso/pull/1463/commits/6938e38b33e4cc0a893a7f95212041dadc62f824#diff-20681b7129ace19298fd33795d95bbf7c5b118c4be2bf33844520b21d8798d45R145-R150) and [having the KB track file contents twice as a `String`](https://github.com/osohq/oso/pull/1463/commits/6938e38b33e4cc0a893a7f95212041dadc62f824#diff-19d5351de096381b72f46df92ca35b5303312561d594ef9b358de9db67ba08deR34) instead of once as a `String` and once as an `Arc<Source>`.